crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -103,7 +103,174 @@ where
     }
 }
 
-type NarrowingConstraints<'db> = FxHashMap<ScopedSymbolId, Type<'db>>;
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub(crate) enum NarrowingType<'db> {
+    Deferred(DeferredType),
+    Eager(Type<'db>),
+    Intersection(NarrowingIntersectionType<'db>),
+    Union(NarrowingUnionType<'db>),
+}
+
+impl<'db> NarrowingType<'db> {
+    /// A simplified version of `NarrowingIntersectionBuilder`.
+    ///
+    /// TODO: Further normalization needed (e.g., ignore if an identical `DeferredType` already exists).
+    fn intersection(
+        db: &'db dyn Db,
+        left: NarrowingType<'db>,
+        right: NarrowingType<'db>,
+    ) -> NarrowingType<'db> {
+        match (left, right) {
+            (NarrowingType::Deferred(deferred_left), NarrowingType::Deferred(deferred_right)) => {
+                deferred_left.intersection(db, deferred_right)
+            }
+            (NarrowingType::Eager(eager_left), NarrowingType::Eager(eager_right)) => {
+                NarrowingType::Eager(
+                    IntersectionBuilder::new(db)
+                        .add_positive(eager_left)
+                        .add_positive(eager_right)
+                        .build(),
+                )
+            }
+            (left, right) => NarrowingType::Intersection(NarrowingIntersectionType::new(
+                db,
+                Box::new([left, right]) as Box<[NarrowingType<'db>]>,
```

Comment:
```suggestion
                Box::from([left, right]),
```

Reply:
Oh !!! Looks more great..! Thanks for fixing it 👍